### PR TITLE
[SPARK-37474][R][DOCS] Migrate SparkR docs to pkgdown

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This PR proposes migration of R API docs to [pkgdown](pkgdown.r-lib.org/). 

Result (not synced automatically) could look like this

https://zero323.gitlab.io/sparkr-docs-experiments/

### Why are the changes needed?

To improve overall experience of interactions with SparkR docs:

- Arguably, much better looking than the simple R docs.
- Fully linked examples allow easy inspection of the used functions.
- Adds functional search.
- For what is worth, all samples can be copied and pasted.
- Dark mode.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Inspection of the build page

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
The demo looks good. cc @falaki @felixcheung @shivaram FYI

File: R/pkg/vignettes/sparkr-vignettes.Rmd
Comment (srowen) [inline R/pkg/vignettes/sparkr-vignettes.Rmd:79]: Out of curiosity why is this needed?

**Review by srowen — COMMENTED** [review]
Seems OK if it gives some benefits and does not entail much change to existing docs

File: R/pkg/vignettes/sparkr-vignettes.Rmd
Comment (zero323) [reply R/pkg/vignettes/sparkr-vignettes.Rmd:79]: `pkgdown` builds vignettes, not only `Rd` files, so `install.spark` fails, because we don't have corresponding release yet.

Comment (zero323) [reply R/pkg/vignettes/sparkr-vignettes.Rmd:79]: It is just a temporary workaround ‒ I'll fix it, once I get it to build outside local environment (I am thinking about adding environment variable that can be set in CI).

File: R/pkg/_pkgdown.yml
Comment (zero323) [inline R/pkg/_pkgdown.yml:26]: Manual reference is not mandatory, but allows us to split index into logical sections, which are easier to digest.

File: R/pkg/_pkgdown.yml
Comment (zero323) [inline R/pkg/_pkgdown.yml:21]: Demo version uses Bootstrap 3, because `preferably` doesn't fully support `pkgdown>1.6` and doesn't work with Bootstrap 5, but such upgrade is an option.

Comment (amirmasoudabdol) [reply R/pkg/_pkgdown.yml:21]: I noticed that this issue is linked to my repo, so I thought I would come here and say that I'm working on an update to support BS5. In the meantime, I just pushed a minor patch that fixes some minor issues from upgrading to `{pkgdown} 2.0`! :-)

Comment (zero323) [reply R/pkg/_pkgdown.yml:21]: Thanks @amirmasoudabdol!

Comment (amirmasoudabdol) [reply R/pkg/_pkgdown.yml:21]: {preferably} now supports {pkgdown} 2.0, and Bootstrap 5! 🎉

It is not yet on CRAN though, I'm waiting for their approval, etc.

Comment (zero323) [reply R/pkg/_pkgdown.yml:21]: Thanks for the update @amirmasoudabdol!

File: R/create-docs.sh
Comment (zero323) [inline R/create-docs.sh:67]: I am not sure if that's the best way to handle different versions, so any suggestions are welcome.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:494]: quick question, do we need this for documentation build? we would have to also update https://github.com/apache/spark/blob/master/docs/README.md and https://github.com/apache/spark/blob/master/dev/create-release/spark-rm/Dockerfile#L83-L84 . Can be done separately though.

And can we pin the versions of both libraries instead of using the master branch?

File: R/pkg/DESCRIPTION
Comment (HyukjinKwon) [inline R/pkg/DESCRIPTION:63]:
```suggestion
RoxygenNote: 7.1.1
```

no biggie but I would keep this diff off.

File: R/pkg/vignettes/sparkr-vignettes.Rmd
Comment (HyukjinKwon) [inline R/pkg/vignettes/sparkr-vignettes.Rmd:31]:
```suggestion
# In GitHub lint job, we don't have full JVM build
```

**Comment by HyukjinKwon** [comment]
Looks pretty good .. but the packaging stuff it would be great to have sign-offs from @felixcheung @shivaram @mengxr or @falaki.

File: .github/workflows/build_and_test.yml
Comment (zero323) [reply .github/workflows/build_and_test.yml:494]: > quick question, do we need this for documentation build?

Yes, documentation build invokes pkgdown. 

> And can we pin the versions of both libraries instead of using the master branch?

Of course. We should be even able to pin CRAN version of `pkgdown` (seems like [2.0.x](https://cloud.r-project.org/web/packages/pkgdown/index.html) and [compatible preferably](https://cloud.r-project.org/web/packages/preferably/))) are already published and visible).

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**
